### PR TITLE
eqrel: Speed up MERGE-EXTEND

### PR DESCRIPTION
**Pull request by langston-barrett**
Copied the iteration pattern from `insertAll`. To be quite honest, I don't 100% understand how this works, but I believe the effect is that we get to skip iterating over any elements that don't actually need to get copied into the delta/new relation.

In my measurements, this takes the runtime of this program down from 6s to 4s, while preserving the result.
```
.decl r(a: number, b: number) eqrel

#define MAX1 2000
#define MAX2 20000

r(1, 1).
r(n+1, n+1) :- r(n, n), n + 1 < MAX1.
r(n, m + (MAX2 * n) + 1) :-
  r(n, m),
  n < MAX1,
  m < MAX2.

.printsize r
```
Will test on my larger program as well.

### CODE REVIEW

File: src/include/souffle/datastructure/EquivalenceRelation.h [hunk 1]
```diff
@@ -172,19 +172,21 @@ class EquivalenceRelation {
...
-        {
-            value_type el;
-            value_type rep;
-            auto it = other.sds.sparseToDenseMap.begin();
-            auto end = other.sds.sparseToDenseMap.end();
-            for (; it != end; ++it) {
-                std::tie(el, std::ignore) = *it;
-                rep = other.sds.findNode(el);
+        other.genAllDisjointSetLists();
+        for (typename StatesMap::chunk it : other.equivalencePartition.getChunks(MAX_THREADS)) {
```
Comment (langston-barrett) [inline src/include/souffle/datastructure/EquivalenceRelation.h:176]: I'm a bit confused by this: the mention of `MAX_THREADS` here (as copied from `insertAll`) seems to indicate that there's some kind of parallelism happening, but it's not like this for-loop starts with `#pragma openmp parallel for` or anything obvious like that... I would appreciate more seasoned C++ programmers thoughts here:

1. Is any parallelism happening here/in `insertAll`? If so, how can you tell?
2. If not, is there any reason to iterate over this in chunks? Should/can I just use the (non-chunked) iterator over `btree`?

**Comment by philzook58** [comment]
Orthogonal to this pull request, out of curiosity, do you have a reference or something I could look at for your unification based pointer analysis?

**Comment by langston-barrett** [comment]
> Orthogonal to this pull request, out of curiosity, do you have a reference or something I could look at for your unification based pointer analysis?

Not yet, but it will be open-source soon!

**Comment by b-scholz** [comment]
For the eqrel data-structures, we would need newly implemented iterators. I did some recent testing with a rewrite system showing still a slow-down using eqrel caused by iterators.

**Comment by b-scholz** [comment]
Could you find any speedups using a parallel version of extend/merge? In the past, this never worked out.

**Comment by langston-barrett** [comment]
> For the eqrel data-structures, we would need newly implemented iterators. I did some recent testing with a rewrite system showing still a slow-down using eqrel caused by iterators.

Is this what's mentioned in #2147? That seems somewhat orthogonal to this MR, but perhaps I'm missing something? I agree that it would be a nice addition.

**Comment by langston-barrett** [comment]
> Could you find any speedups using a parallel version of extend/merge? In the past, this never worked out.

I'm not sure I'm the best person to try this out - as indicated in a comment above, I'm already a tad confused about the parallelism (or lack thereof) in this code. Generally, I have no experience with parallel C++.

**Comment by b-scholz** [comment]
We have a very simple parallelisation strategy. Only the first outer loop is parallelised but it must be a complex loop (at least two nested loops). With simple loops such as merging we could never find a speed-up but of course it would be great to revisit the parallelisation strategy.

File: src/include/souffle/datastructure/EquivalenceRelation.h (hunk 1 above)
Comment (OlivierHamel) [reply src/include/souffle/datastructure/EquivalenceRelation.h:176]: Looking at the implementation of `getChunks`, it looks like it does exactly what it sounds like: It iterates over the set, and splits it into `n` sequences of length approx. `|C| / n`.

Looks like a copy-paste mistake. I'd say remove it entirely, both here and in `insertAll`, and just iterate over the collection.

**Comment by langston-barrett** [comment]
Will reopen if/when I have some time to revisit this.

**Comment by b-scholz** [comment]
Can we link this issue with the previous pull-request (and believe there might be also an older issue) about eqrel merge? There is a lot of explanation / info there.

**Comment by langston-barrett** [comment]
> Orthogonal to this pull request, out of curiosity, do you have a reference or something I could look at for your unification based pointer analysis?

@philzook58 https://galoisinc.github.io/cclyzerpp/implementation.html#unification-based-analysis-via-subsumption-and-choice

**Pull request closed without merging**